components/shared/net/request.rs

```diff
@@ -703,6 +706,27 @@ impl Request {
         self.url_list.first().unwrap().clone()
     }
 
+    pub fn original_url(&self) -> ServoUrl {
+        match self.mode {
+            RequestMode::WebSocket {
+                protocols: _,
+                ref original_url,
+            } => original_url.clone(),
+            _ => self.url(),
+        }
+    }
+
+    pub fn original_url_mut(&mut self) -> &mut ServoUrl {
+        match self.mode {
+            RequestMode::WebSocket {
+                protocols: _,
+                ref mut original_url,
+            } => original_url,
+            // FIXME(pylbrecht): do we want url_mut() instead?
+            _ => self.url_list.first_mut().unwrap(),
```

Comment:
I am not sure if we want to have `url_mut()` on `Request`, as the URL seems to be the request's identifier. Mutating that would sort of defy its purpose, wouldn't it?